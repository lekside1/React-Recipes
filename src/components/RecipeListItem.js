import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

const RecipeListItem = ({
  recipe,
  favorited,
  onClick,
  onFavorited,
}) => (
  <li
    key={recipe.id}
    className="py2 border-bottom border-bottom-dashed pointer"
    onClick={() => onClick(recipe.id)}
  >
    <span
      className="mr1"
      onClick={e => {
        e.stopPropagation();
        onFavorited(recipe.id);
      }}
      role="img"
      aria-label="favorite"
    >
      {favorited ? '✅' : '⬜️'}
    </span>
    <span>{recipe.name} </span>
    <span>{recipe.category}</span>
    <span>
      {favorited
        ? (
          <Link to={`/recipe/${recipe.id}`} className="text-decoration-none">
            <span
              className="ml1"
              onClick={() => onClick(recipe.id)}
              role="img"
              aria-label="detail"
            >
              ➕
            </span>
          </Link>
        )
        : ''}
    </span>
  </li>
);

RecipeListItem.propTypes = {
  recipe: PropTypes.object,
  favorited: PropTypes.bool,
  onClick: PropTypes.func,
  onFavorited: PropTypes.func,
};

RecipeListItem.defaultProps = {
  recipe: {},
  onClick: () => {},
  onFavorited: () => {},
};

export default RecipeListItem;
