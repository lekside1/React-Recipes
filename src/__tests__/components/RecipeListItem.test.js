import React from 'react';
import { mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import RecipeListItem from '../../components/RecipeListItem';

const testRecipe = {
  id: 1,
  name: 'Test recipe',
  category: 'Dessert',
};

describe('<RecipeListItem />', () => {
  let component;
  let tree;

  test('Should not break if no recipe passed', () => {
    expect(() => renderer.create(
      <BrowserRouter>
        <RecipeListItem />)).not.toThrow();
      </BrowserRouter>,
    ));
  });

  test('Should render passed recipe', () => {
    component = renderer.create(
      <BrowserRouter>
        <RecipeListItem recipe={testRecipe} />);
      </BrowserRouter>,
    );
    tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('Should render favorited state', () => {
    component = renderer.create(
      <BrowserRouter>
        <RecipeListItem recipe={testRecipe} favorited />,
      </BrowserRouter>,
    );
    tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('Should call onClick when clicked', () => {
    const onClick = jest.fn();

    component = mount(
      <BrowserRouter>
        <RecipeListItem recipe={testRecipe} onClick={onClick} />);
      </BrowserRouter>,
    );
    component.simulate('click');

    expect(onClick.mock.calls.length).toBe(1);
    expect(onClick.mock.calls[0][0]).toBe(testRecipe.id);
  });

  test('Should call onFavorited when favorited', () => {
    const onFavorited = jest.fn();

    component = mount(
      <BrowserRouter>
        <RecipeListItem recipe={testRecipe} onFavorited={onFavorited} />,
      </BrowserRouter>,
    );
    component.find('div[role="button"]').simulate('click');

    expect(onFavorited.mock.calls.length).toBe(1);
    expect(onFavorited.mock.calls[0][0]).toBe(testRecipe.id);
  });

  // test('Should not call onClick when favorited', () => {
  //   const onClick = jest.fn();
  //   const onFavorited = jest.fn();

  //   component = mount(
  //     <BrowserRouter>
  //       <RecipeListItem
  //         recipe={testRecipe}
  //         onClick={onClick}
  //         onFavorited={onFavorited}
  //       />
  //     </BrowserRouter>,
  //   );
  //   component.find('div[role="button"]').simulate('click');

  //   expect(onFavorited.mock.calls.length).toBe(1);
  //   expect(onClick.mock.calls.length).toBe(0);
  // });
});
