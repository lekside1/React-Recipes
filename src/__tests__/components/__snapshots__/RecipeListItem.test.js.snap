// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecipeListItem /> Should render favorited state 1`] = `
Array [
  <li
    className="py2 border-bottom border-bottom-dashed pointer nowrap"
    onClick={[Function]}
  >
    <div
      className="inline-block animated mr1"
      onClick={[Function]}
      role="button"
    >
      ✅
    </div>
    <span>
      Test recipe
       
    </span>
    <span>
      Dessert
    </span>
    <span>
      <a
        className="text-decoration-none"
        href="/recipe/1"
        onClick={[Function]}
      >
        <span
          aria-label="detail"
          className="ml1"
          onClick={[Function]}
          role="img"
        >
          ➕
        </span>
      </a>
    </span>
  </li>,
  ",",
]
`;

exports[`<RecipeListItem /> Should render passed recipe 1`] = `
Array [
  <li
    className="py2 border-bottom border-bottom-dashed pointer nowrap"
    onClick={[Function]}
  >
    <div
      className="inline-block animated mr1"
      onClick={[Function]}
      role="button"
    >
      ⬜️
    </div>
    <span>
      Test recipe
       
    </span>
    <span>
      Dessert
    </span>
    <span>
      
    </span>
  </li>,
  ");",
]
`;
